types/types.go

```diff
@@ -148,86 +148,66 @@ func (m *memMarker) Count(states ...AlertState) int {
 }
 
 // SetSilenced implements Marker.
-func (m *memMarker) SetSilenced(alert model.Fingerprint, version int, ids ...string) {
+func (m *memMarker) SetSilenced(alert model.Fingerprint, version int, activeIDs []string, pendingIDs []string) {
 	m.mtx.Lock()
+	defer m.mtx.Unlock()
 
 	s, found := m.m[alert]
 	if !found {
 		s = &AlertStatus{}
 		m.m[alert] = s
 	}
+	s.SilencedBy = activeIDs
+	s.pendingSilences = pendingIDs
 	s.silencesVersion = version
 
 	// If there are any silence or alert IDs associated with the
 	// fingerprint, it is suppressed. Otherwise, set it to
-	// AlertStateUnprocessed.
-	if len(ids) == 0 && len(s.InhibitedBy) == 0 {
-		m.mtx.Unlock()
-		m.SetActive(alert)
+	// AlertStateActive.
+	if len(activeIDs) == 0 && len(s.InhibitedBy) == 0 {
+		s.State = AlertStateActive
 		return
 	}
 
 	s.State = AlertStateSuppressed
-	s.SilencedBy = ids
-
-	m.mtx.Unlock()
 }
 
 // SetInhibited implements Marker.
 func (m *memMarker) SetInhibited(alert model.Fingerprint, ids ...string) {
 	m.mtx.Lock()
+	defer m.mtx.Unlock()
 
 	s, found := m.m[alert]
 	if !found {
 		s = &AlertStatus{}
 		m.m[alert] = s
 	}
```

Comment:
I think that wouldn't work because `m.Status()` would try to acquire the mutex again.

Reply:
A common pattern in Prometheus (which I tend to like) is that we do:

```
// Status implements Marker.
func (m *memMarker) Status(alert model.Fingerprint) AlertStatus {
	m.mtx.RLock()
	defer m.mtx.RUnlock()

	return m.status(alert)
}

func (m *memMarker) status(alert model.Fingerprint) AlertStatus {
	if s, found := m.m[alert]; found {
		return *s
	}
	return AlertStatus{
		State:       AlertStateUnprocessed,
		SilencedBy:  []string{},
		InhibitedBy: []string{},
	}
}
```

the suggestion is a bit silly but I find it quite important as it can be easy to miss when changing the signature of `Status`, and help reduce [shotgun surgery](https://en.wikipedia.org/wiki/Shotgun_surgery#:~:text=Shotgun%20surgery%20is%20an%20antipattern,implementations%20in%20a%20single%20change).

Reply:
Yes, that's a good point.

However, I would do such a refactoring separate from this PR as it is not really related.

Also, note that the `Status` method does something slightly different: It just returns an "empty" `AlertStatus` for a fingerprint that doesn't exist, _without adding it to the map_. It also returns a copy of the `AlertStatus` rather than a pointer. `SetInhibited` and `SetSilenced`, however, will modify the returned AlertStatus, so they need a pointer, and they need to add a newly created `AlertStatus` to the map.

I guess all of this can be somehow solved, but since it's getting pretty involved at this point, I am a bit concerned that we would break some assumptions we might not be aware of right now. I'm not super familiar with the code base after all. That's another reason to not do such a refactoring at the same time as a bugfix.

Reply:
Perfectly sensible - let's defer for now.